File: 34_decimal_to_binary/34_decimal_to_binary.py
# Python program to convert decimal to binary

# method 1 - using brut-force(mind and sense) and iteration
try:
    n=int(input("Enter the decimal number :"))
    string=""
    while n>0:
        a=n%2
        n=n//2
        string=string+str(a)
    print((string)[::-1])
except:
    print("Enter the valid integer value")

# method 2 - using recursion
try:
    n=int(input("Enter the number :"))
    srt=""
    def func(n):
        if n<=0:
            return print(end="")
        s=n%2
        global srt
        srt=srt+str(s)
        func(n//2)
    func(n)
    print(srt[::-1])
except:
    print("Enter the valid integer value")

# method 3 - using built-in method
try:
    n=int(input("Enter the number :"))
    print(bin(n))
except:
    print("ENter a valid interger vaLUE")
    
# method 4 - using module based approach

def decimal_to_binary(decimal):
    binary_num = ""
    while decimal > 0:
        remainder = decimal % 2
        decimal = decimal // 2
        binary_num = binary_num + str(remainder)
    return binary_num[::-1]

try:
    while (True):
        num = int(input("Enter positive integer value : "))

        if (num < 0):
            print("Enter valid decimal number")
        else:
            result = decimal_to_binary(num)
            print(f"Decimal {num} = {result} Binary")

        choice = input("Do you want to continue the program (y/n) : ")
        if(choice.lower() == "n"):
            print("Program finished!")
            break

except ValueError as e:
    print(f"Invalid Input : {e}")

except ZeroDivisionError as e:
    print(f"Zero division exception : {e}")

except Exception as e:
    print(f"Exception caught : {e}")

File: 34_decimal_to_binary/34_decimal_to_binary.dart
import 'dart:io';

// program to convert decomal number into binary format

String reverseString(String num) {
  String revString = "";
  for (int i = revString.length - 1; i <= 0; i--) {
    revString = revString + num[i];
  }
  return revString;
}

String decimalToBinary(int decimal) {
  String binaryNum = "";
  while (decimal > 0) {
    int remainder = decimal % 2;
    decimal = decimal ~/ 2;
    binaryNum = binaryNum + remainder.toString();
  }
  return reverseString(binaryNum);
}

void main() {
  try {
    while (true) {
      stdout.write("Enter a positive integer value : ");
      int num = int.parse(stdin.readLineSync()!);

      if (num < 0) {
        print("Enter the value in decimal");
      } else {
        String result = decimalToBinary(num);
        print("Decimal $num = $result Binary");
      }
      stdout.write("Do you want to continue the program (y/n) : ");
      String choice = stdin.readLineSync()!.toLowerCase();
      if (choice == 'n') {
        print("Program finished!");
        break;
      }
    }
  } on FormatException catch (e) {
    print("Invalid input : $e");
  } on UnsupportedError catch (e) {
    print("Divide by zero exception : $e");
  } on Exception catch (e) {
    print("Exception caught : $e");
  }
}
